### DISCUSSION TITLE
Multi-user / permission support

### DISCUSSION THREAD

**Discussion by deajan** [TriliumNext/Trilium#5899] [General]
So basically I wrote about this earlier in zadam's post about trilium's maintenance status.

This has been one of the major feature quite a lot of people asked for.
Making trilium multi-user friendly, eg:
- Add a permission system for notes (ie CRUD permissions)
- Add user authentiication in order to get permission on a per user basis
- Optionally add groups

The major benefit I see here is the traction that TriliumNext would get, ie it would be used in enterprise where people usually document with wiki like systems (ex docuwiki).

Someone stated in the original issue that Trilium is a _PERSONAL knowledge base_, but IMO tadding optional permissions wouldn't bother people who like to use it that way.
Still il would allow sharing the same knowledge base accross multiple users, personal or professionnal wise.

From a developper point of view, I think the major problem here is that notes hierarchy isn't fixed, and notes can be cloned, so the most complicated part would be to make permissions inheritance on notes that have multiple parents.
This could perhaps be solved by having a note owner, which is the only user which permissions are being inherited.
That of course would also need a way to change the ownership of the notes.

Anyway, I am really hoping this idea to get picked up, since it would probably mean way more people attracted to the project.

**Comment by zerebos** [comment]
I think this in general is a very good idea to include. As you said, it's probably the most asked for feature from what I've seen. But even from what you've described, there's multiple ways this can be handled. We can have multiple users using the same instance but completely different databases, or we can have them split up but within the same database, or we can do as you suggest with a full permission system that would have users potentially working in the same hierarchies.

There's a lot to think about here. The simplest one to do is have multiple users on the same "instance" but using separate databases. That would still allow it to be a personal database rather than a collaboration platform that has users on the same hierarchies.

**Reply by meichthys** [reply]
An even simpler implementation may be to just allow password protecting of individual notes using a password other than the instance password. This way multiple users could share the instance password, and then still define their own passwords on portions of the hierarchy. This would allow limited collaboration (not real-time of course).

**Reply by edgeTurbo** [reply]
@meichthys I want to implement this feature: having multiple users on the same "instance" but using separate databases, but I'm at a loss. Can you give me some advice,thanks!

**Reply by edgeTurbo** [reply]
@meichthys Just provide an idea and related code, I really don't know where to start, thank you very much

**Reply by edgeTurbo** [reply]
@meichthys I'm not sure if this is feasible. Is it possible to declare the SQLite database file using sql.ts and pass a dynamic parameter?
But I found that this sql.ts only runs once when running the server. How to modify it to dynamically call different SQLite files? I am not very familiar with front-end knowledge, thank you very much!
<img width="1169" alt="image" src="https://github.com/user-attachments/assets/b1d65d6d-80ed-476a-9e6d-854d199d905c">

**Reply by meichthys** [reply]
@eliandoran could you point @edgeTurbo in the right direction?

**Reply by eliandoran** [reply]
@edgeTurbo , can you give me more info on what you're trying to achieve? I understand that you would like to add support for having different databases, but do you want to do it only once at start-up, to be able to switch between them like you would do a File -> Open in text editors, or do you want to be able to use multiple databases at the same time? 

Currently the whole architecture is done using a single database that is initalized at start-up.

If you want to achieve some kind of basic multi-user support then you need to do it securely so you can't simply specify which database to use from client to server. Most likely you'd have to store that information in the login user session (see `session_parser`) and to keep a database in memory for every logged in user.

The most difficult part is replacing the `sql` singleton (which is used almost everywhere) with something that is aware of the session.

Since you're dealing with user data, it's also very important to take into account security. It's inadmissible to have leaks between DBs.

Regardless I believe this is a too heavy endevour for now. I would suggest taking some time doing some smaller tasks to get accustomed to the codebase and maybe then you could tackle this big subject.

**Reply by edgeTurbo** [reply]
@eliandoran yes,My goal is to have different users carry token requests, and the token is determined by redis or database to determine which user it is and which path the user's sqlite file is in. After obtaining this path, new Database(dataDirPath) is used each time when performing CRUD operations of better-sqlite3, just like opening and switching files in a text editor. For some reason, I need to complete this multi-user functional demo as soon as possible, and I don't need to consider some deeper issues for the time being, so I would like to ask if my approach is correct? However, because the entire architecture is completed using a single database initialized at startup, is it possible not to specify a database at startup, and when the user logs in with a token, load different sqlite files through the database to obtain the user's data? I don't know how to modify this yet, can you give me some valuable suggestions? Thank you very much!

**Reply by edgeTurbo** [reply]
@eliandoran Indeed, as you said, I have updated the user information into the session, but now there is a problem that is bothering me. I am unable to globally obtain the session corresponding to the request. I want to make the smallest possible changes, so I want to use the global session retrieval method in sql.ts to obtain the dataDir of the corresponding user, and then dynamically change the database file in sql.ts. But I don't know how to obtain the session globally. It seems that it can only be obtained through routing. I am not very familiar with Node.js. If there is a solution, please let me know. Thank you!

**Reply by edgeTurbo** [reply]
<img width="1837" alt="image" src="https://github.com/user-attachments/assets/dc8535ad-b567-40a2-90ec-dd796c591447">
<img width="1010" alt="image" src="https://github.com/user-attachments/assets/d0231770-8f6f-4d9b-a3ce-c4dff5be8136">

**Reply by eliandoran** [reply]
@edgeTurbo , you could try to hook into `auth.ts` in the `checkApiAuth`. This method appears to be called before every API request. You there have access to the `req.sessions` and you can then call `sql` to tell it to change database.

**Reply by edgeTurbo** [reply]
@eliandoran Thank you very much for your suggestion. I am able to continue writing the code. Indeed, when requesting interfaces, I always go to the checkApiAuth method. Now, I am using cls-hooked to retrieve the corresponding user's database in sql.ts.Here are some of the code modifications I made
<img width="1126" alt="image" src="https://github.com/user-attachments/assets/eb690adf-7727-4383-827c-56d80f647cca">
<img width="1021" alt="image" src="https://github.com/user-attachments/assets/0c6d4bf7-9b4f-48db-9795-5dfa6a48ce06">
<img width="1216" alt="image" src="https://github.com/user-attachments/assets/379d7d75-b6d8-4804-a275-c601d78b2ce0">
It can indeed produce some effects, but unfortunately, when the service starts, it will perform sql_init.initializeDb() and write it to the Rebecca object. After I block sql_init.initializeDb(), I will get an error Error: get /api/tree thread exception: 'Note' root 'not found in the cache', stack: undefined. The reason should be that the becca object does not exist. After login is successful, it will jump to '/' route. When rendering in index.ts, the database needs to be queried, but there is no user database instance in the namespace at this time. Therefore, when the service starts, some queries will go to the original database instance. Only when a request is made. When using the API, it will go to checkApiauth to generate a database instance. How should I solve this? I am very sorry, my coding skills are very beginner, and I hope you can give me more help

<img width="1319" alt="image" src="https://github.com/user-attachments/assets/1226e6ea-ea9f-44d0-a867-ebb7389c417b">

**Reply by edgeTurbo** [reply]
@eliandoran I found that when the service starts, the becca object is initialized. When obtaining notes or trees, it is obtained from becca object. However, different databases obtain the same becca object because the app. ts initializeDb func, and at this time, the database instance is still initialized through  The database instantiated by 'dataDir.DOCUMENT_PATH'. So when I switch to a new database for querying the information in the DOCUMENT_PATH of the becca object, I will retrieve the note ID from becca to request the database, and at this time, an error will be reported stating that the note_id is not in the database. I don't know if it's the situation I described, could you please provide me with more help or ideas

**Comment by zadam** [comment]
I'll add my 2 cents.

I'd start by gathering what use cases would Trilium Next want to target. "multi-user" or "permissions" are very generic terms which can span from very simple multi-tenancy to extremely complicated systems. It would be good to find a good balance of the most useful use cases covered, fit to the existing Trilium features / philosophy, and the implementation cost needed.

The other thought I have is this - it looks like the new org highest priority is refactoring. I'd recommend focusing on that first before tackling such major features as multi-user. It's good of course to think about the future, but not diverting attention into 10 different directions is IMHO an important consideration too.

**Reply by zerebos** [reply]
> I'd start by gathering what use cases would Trilium Next want to target. 

Obviously I can't speak for everyone involved, but I think the ability for multiple to use the same installation of Trilium and each have their own personal knowledge database fits the Trilium philosophy the best :-)

> It's good of course to think about the future, but not diverting attention into 10 different directions

I completely agree with you on this! Personally, I just wanted to have these kinds of major changes in the back of my mind while refactoring because I might consciously or unconsciously notice key pieces of code or even architecture that are relevant to the major changes that I can use later on when those problems are actually tackled.

**Reply by rmkni** [reply]
> it looks like the new org highest priority is refactoring

could you link discussions about this?

**Reply by alrhalford** [reply]
For what it's worth, I don't think full multi-user stuff is useful. Being able to share a specific note(tree) as editable, maybe, but not fully fledged notion style support. I would be supportive of being able to export and import trees to/from some text format though, that would be nice.

**Reply by meichthys** [reply]
> > it looks like the new org highest priority is refactoring
> 
> could you link discussions about this?

@rmkni See here:
https://github.com/orgs/TriliumNext/discussions/26

https://github.com/orgs/TriliumNext/discussions/4

**Reply by meichthys** [reply]
> I would be supportive of being able to export and import trees to/from some text format though, that would be nice.

I think we can already do this with the normal import/export feature, no?

**Reply by alrhalford** [reply]
Quite possibly, I haven't used that functionality - apologies if it is!

**Reply by maphew** [reply]
Looking for shared trees is what brought me to this thread. My use case is: the hub of my star topology is at home on my NAS, with spokes being desktop, laptop, phone, and work computers. Where I want to be is that, plus: an Trilium server that has only certain branches of my database. These branches would be used as Shared, giving me a website of my knowledgebase but not risking exposing everything should the public server be compromised. 

I think this scenario could dovetail nicely with multi-user. Imagine certain branches being tagged `#sync:https/hostname.net/path/` and automatically synchronized using the regular mechanism. Each sync host manages their own permissions locally no different than now, with one new password / authentication mechanism just for the sync api.

(Thinking more broadly, the destination for this imagined `#syncTarget` tag need not be restricted to http. e.g. it could be another sqlite db on disk, or even a 'real' rdbms like postgres, an rsync daemon, or scm like fossil or git.)

That said, as much as I would like this _tomorrow!_ I agree the current refactoring effort should have priority.

**Reply by Phalen** [reply]
First of all I must state this BLOWS onenote out of the water. This is every thing i wanted out of onenote and so much more....  THANKYOU!!!  

As for the multi user aspect you first have to decide 3 things.  ;)

1 are you going to replace onenote or sharepoint in style.  
2. should each root be its own database and have its own password/security kinda like how onenote can open multiple notebooks/ share point you have your "my site"  and then you can also attach/map team sites, and even some one elses "my site" or one large conglomerate database thing that needs mysql or other larger databases to handle the security and different style of access.
3. do you plan to eventually allow/make it big enough for an enterprise which partially falls under TriliumNext/Notes#2 for what type of database style. 

I can see this product be loved by entire households as well as smb and possibly medium businesses pretty much any which way you take it. 

I could see this becoming a offline capable sharepoint honestly. If you figure out the right way to do the front end i could seriously see it being used every where. 

Since it is refactoring time i would suggest determining at least the general frame work you want to use to go multi user so during the refactor you can see what parts need adjustments to handle multi users and possibly even lay ground work for when you start coding in the user support.

**Comment by eliandoran** [comment]
@sigaloid , as developer of trilium.cc, what's your take on this?
Without going in too much detail, how does trilium.cc achieve multi-user support?
Is there something we can implement in the server to ease multi-user support?

**Reply by sigaloid** [reply]
Unfortunately, I don't do any real multi-user support - I simply dockerize each server and run each one separately (with separated data volumes).

I wouldn't want to have one server with multiple users anyway - the containerization is very helpful in terms of security (a vulnerability in the code base isn't remote shell - just a docker container).

side note: this is also why I've been researching alternative runtimes like deno or bun - I pay the Node runtime cost for every single container which means compute is far more than ideal. Also why I've looked into writing a Rust backend as a sync target.

**Comment by deajan** [comment]
Some kind of multi-tenanncy can be be achieved by running multiple trilium instances with different trilium-data dirs.
Realtime collaboration is another beast, which is FAR from the most needed I guess.

@rauenzi Having the same database for multiple users would be the best IMO.
The usecase I'm targetting here is sharing notes with family / collegues, for them to see and/or edit, depeding on permissions.

**Reply by meichthys** [reply]
> Having the same database for multiple users would be the best IMO.\nThe usecase I'm targetting here is sharing notes with family / collegues, for them to see and/or edit, depeding on permissions

Yes. This seems to be the most requested use case from what ice seen in past issues.

**Reply by alrhalford** [reply]
I would be interested to know the extent to which people want to collaborate on an entire instance. A couple of thoughts:

The ability to share an authenticated link that gives the recipient r or r/w access to a particular note tree, a bit like google docs does. The downside of this is that it becomes difficult to say "they can read this note tree, but only edit this subtree, etc", but the upside is that it should be easy to implement, and permissions are easy (if not super expressive).

The idea of "remote notes" in your own trilium instance that reference a note tree that is stored elsewhere. The downside is that you need your own trilium instance, but the upside is that it slots in quite nicely without re-engineering the existing stuff too much.

**Reply by deajan** [reply]
As of today, links to note trees are a simple HTML representation of the notes, without any editor function.
Having an authenticated link like you suggest would mean for a foreign user to get access to the whole Trilium interface, with all the frontend JS/plugins, which would probably lead to major security issues.

Also, "remote notes" could simply disappear without further notice, which is also not a good practice in terms of note taking.

From what I understand, the major multiuser need here is to have multiple users collaborate on the same Trilium instance.
Each user could have private notes, and view/edit group notes.
@meichthys What's your take on this ?

**Reply by alrhalford** [reply]
To be clear, I personally don't see the value in multi user, but to address your points:

Authenticated links: Clearly the authenticated link would not work now, work would need to be done to allow such links to be scoped to particular note trees, and the backend would only return the note trees that are within scope.

Remote notes: you're right, somebody could delete a remote note and it would disappear, but this is not so different from normal collaboration tools - if you and I collaborate on a Google doc/notion etc, I could just wipe/delete a page and it would disappear for you (local backups, versioning, etc notwithstanding).

**Reply by meichthys** [reply]
@deajan I do agree that multi user had significant benefits in cases like a small office, or family scenarios where there are multiple people who are expected to have read write access on a regular basis.
I have a deployment in a small office and we currently only have one way to restrict access to certain note trees (password protecting with the instance password). We logged in each user with the instance password (but never gave the password to them h this way we can 'reuse' the password to lock down 'IT or sensitive sections' of the notebook.

Some multi user discussions happened here: https://github.com/zadam/trilium/discussions/3921 <–– see this discussion for a list of old issues requesting this kind of feature
https://github.com/zadam/trilium/discussions/1375
https://github.com/zadam/trilium/discussions/3024

To me it breaks down to two categories: users that want to be able to just have two databases on the same instance (multi tenancy), and users who want some kind of better collaboration (whatever collaboration means). 
A read\write authenticated link would satisfy a portion of the use cases, but not most from what I've seen in previous requests for these kinds of features.

**Comment by GitTimeraider** [comment]
Can add a vote for this. 
But I also have to agree that unless some major parts get overhauled, I feel like most people would be more than fine already with multiple people having acces to the same database data and a very basic permission system (even if it does start off with having to literally tag everything with specific usernames)

More advanced stuff can be kept in mind during development but might not need to be a primary focus.
Not sure how many people are going as far as to suggest "collaboration" options, but not sure whether its to great of an idea to get to far into that.
Im all for more features and if stuff is possible it is nice, but I feel like full-on collaboration projects and notes simply work better in dedicated software so not sure whether that will lead to feeling like to much is crammed into 1 packages while not all of it will be as fleshed out.

**Comment by Br0kenSilos** [comment]
My use case its pretty simple although the implementation may or may not be. The ability to have multiple family members login using their own passwords. Once logged in, permissions (ie Master, Edit, Read-only, Share) could define their access levels to individual branches of the main tree only and propagate down through the sub-branches. A Primary "Master" would have full access to the whole tree. Then the individuals could further control their own branch downward. So For instance..

ROOT (Master)
    \Dad's Notes (Dad)
    \Mom's Notes (Mom)
    \Kid 1 Notes (Kid 1)
    \Kid 2 Notes (Kid 2)
    \Family Shared (All Members)

**Reply by deajan** [reply]
@Br0kenSilos Sounds to be a pretty good usecase. Add "parents" and "children" groups to the permissions, and I'm all in ;)

**Comment by CrO2Cl2** [comment]
to add something: I think High Availability is a must for multi user support, in case there are any very big instances. You can already "chain" multiple servers together, but it would be nice to be able to make  "web" of servers instead of just a chain
(issue that asked for HA: zadam/trilium#2654 )

**Reply by dyllan-to-you** [reply]
To make high-availability easy could utilize [hiqlite](https://github.com/sebadob/hiqlite)

**Comment by 0xStuart** [comment]
I have been using DekiWiki for nearly 20 years before it was commercialised  into MindTouch.  Until now, I have been unable to find anything to replace it with.

Trilium is DekiWiki without the user privileges.  It's the simplicity of DekiWiki, that made it very useful. You could re-skin it to make web sites, and hide the login page. It would default to read-only, and became very popular with web developers.

All Trilium needs, is the ability to create users, and groups, with read/write attributes which can be attached to documents.

**Reply by maphew** [reply]
Another deki-user! It/Mindtouch was the only notes system which worked equally well for me at home as well as work. I poured _so_ much energy into adapting and using it. The burn from them abandoning their established customer base on pursuit of 'enterprise' still hurts. And it's not like I wasn't willing to pay for both deployments. Unlike you I didn't stay with it. I came in at the early Mindtouch transition and only had deki for a few months. Onenote is what I ended up settling in the intervening fifteen years.

I echo your opinion that Trilium is the only thing I've found that could be a suitable replacement.

**Reply by meichthys** [reply]
As per our latest community meeting and discussion, i gather that multi-user or multi-tenant use cases are relatively high on the priority list (once the app and delivery pipelines are stable and testable.)

**Comment by glaucon1984** [comment]
I think it would be easier to implement a system where multiple users can have a DB in a single server, and leave collaboration for the future.
It's already possible to export/import a tree of notes and even sharing through URL.

If there was a minimum level of LDAP integration this could be a great tool in the enterprise. There are already collaboration and sharing tools like SharePoint and OneNote, that most companies use, but those tools are slow and hard to use when one just wants to keep some notes. And you end up with people having hundreds of untitled notes in Notepad++ that are impossible to organize or search.

As stated, it's a personal knowledge base, and I'd really love to put it in my workplace, so each worker can have their personal knowledge base without concerns for sensitive data being posted in some random cloud service.

But I cannot deploy and manage one instance per user.

Give me multi-user with LDAP integration so I can rely on AD authentication and group membership for the access, and I would start immediately doing the paperwork to get the app approved in my workplace.

The behaviour of the Windows application sync'ing with the web server is perfect for our use case, users can run the Windows app in their company laptop and have access to their notes even offline, and as soon as VPN is available then they can sync with the server, which we can make sure is properly backed up.

Also, our security team will be very happy to approve an open source project that will always keep the data within the company infrastructure.

**Comment by dyllan-to-you** [comment]
Crossposting https://github.com/TriliumNext/Notes/issues/56 for reference

I'm uh not a Trillium user yet, but someone on Hacker News has convinced me to take a look wrt migrating off of LogSeq. So that's on my TODO list. That said there's a lot of interest in something that maintains data sovereignty at my work.

For HighAvailability capabilities like someone asked for earlier, why not use [hiqlite](https://github.com/sebadob/hiqlite)? Although I guess a question I have is how sync is handled Currently. 

For multi-user/authentication and permissions/authorization, why not take advantage of [Ory Kratos](https://github.com/ory/kratos) and [Ory Keto](https://github.com/ory/keto)? Free, Open Source, Commercially Backed products. Then the work just becomes a matter of integration (and might help spur some sponsorships 😜). 

As for the general model, here's what I'm kinda imagining as a path. With the caveat that I haven't dug into Trillium's code or architecture yet and am just working off my understanding of the documentation and general system design knowledge.

# Milestone 1: Multi-User
- Support in the Trillium Client for Multiple Local Databases. For an individual, this would mainly be useful if I wanted to have a hard separation between say my personal life notes and my personal work notes
- Update things such that sync servers are configured per-database. That will get you some basic collaboration, without auth. 
- Add Ory Kratos support/integration to the sync server, and update the client to support this kind of auth. 
# Milestone 2: Permissions
- Figure out a nice permissions model that makes sense, ideally something that fits [ory keto](https://github.com/ory/keto)/google zanzibar
  - Personally I like the idea of Owner/Admin/Editor/Viewer, where Owner owns the whole thing and can assign other roles, Admin can be assigned to the whole DB or just a note and has R/W and permissions for all sub-notes, Editor has R/W but no permissions modification access, and Viewer is just ReadOnly. A role's permission cascades down the entire tree, and includes cloned notes. 
- Keto<->Trillium integration
# Milestone 3: Live Sync
- something something CRDTs. Gotta dig into the database a bit more to propose something more coherent
# Milestone 4: "Merge Requests"
- my personal pet feature desire :D

Still a bunch of work, but using existing open source projects should make life easier! Whenever I get around to migrating off of LogSeq, and if I find myself happy with Trillium, I'd be happy to contribute to help make this happen. 

EDIT: _oooooor_ if my understanding of how things work lines up how I imagine, [Electric-SQL](https://electric-sql.com/) might work wonders for both high availability, synchronization, and collaboration. It says postgres everywhere but i think it might work with a sqlite clientdb. Otherwise, I guess I'd need to dig into how the db connection is structured and queries are made and what the schema even looks like in order to get a clear idea on how much work it would be to migrate to postgresql and pglite from sqlite. It should handle a decent chunk of the live sync/collaboration features (but probably not the hardest part; the body of a note).

**Comment by TheFreeman** [comment]
I have opened a focused design proposal based on the use cases discussed here:

- every user retains a personal note tree;
- selected branches can be shared as read-only or editable;
- the sharing user remains the owner;
- permissions are inherited through the shared branch;
- no team workspaces, organization hierarchy, enterprise permission matrix, or real-time collaboration are introduced.

The proposal is intended to preserve Trilium’s personal knowledge-base model while adding controlled collaboration.

Design proposal:
https://github.com/orgs/TriliumNext/discussions/10687

**Reply by deajan** [reply]
@TheFreeman We currently have a write up at https://github.com/TriliumNext/Trilium/issues/4956#issuecomment-3572299547 and a pending PR at https://github.com/TriliumNext/Trilium/pull/10348 open

Perhaps it's a bit late to restart from scratch, but you're welcome to review my writeup and @SashankBhamidi's pull requests
